### ISSUE TITLE
Segmentation fault with using Jderobot Image library with ROS2

### ISSUE THREAD

**Issue by pankhurivanjani** [TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani#7]
This issue is coming in CamViz tool when testing witth ROS2. 

* runtime error

* header file of jderobot which has this error: /opt/jderobot/include/jderobot/types/image.h

* Possible reasons:

    1. Gcc/g++ version error in compiling opencv 
    
     2. To be precise, this error is coming initializing an empty matrix in a class declaration.

     3. Same header file(from jderobot library) is working fine with ROS1. 

Exact error after debugging:

``` 
pankhuri@pankhuri-LENOVO-IDEAPAD-500-15ISK:~/week8-camviz/build/camviz$ gdb ./camViz camViz.yml 
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./camViz...done.
"/home/pankhuri/week8-camviz/build/camviz/camViz.yml" is not a core dump: File format not recognized
(gdb) run
Starting program: /home/pankhuri/week8-camviz/build/camviz/camViz 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffd34af700 (LWP 21297)]
[New Thread 0x7fffd2cae700 (LWP 21298)]
Loading glade
hoReceiving ROS2 messages
[New Thread 0x7fffcb895700 (LWP 21299)]
[New Thread 0x7fffcb094700 (LWP 21300)]
[New Thread 0x7fffca893700 (LWP 21301)]
[New Thread 0x7fffca092700 (LWP 21302)]
[New Thread 0x7fffc9891700 (LWP 21303)]
[New Thread 0x7fffc9090700 (LWP 21304)]
[INFO] [camViz]: constructor Called
[New Thread 0x7fffb7fff700 (LWP 21305)]
[New Thread 0x7fffb77fe700 (LWP 21306)]
[New Thread 0x7fffb6ffd700 (LWP 21307)]
[New Thread 0x7fffb67fc700 (LWP 21308)]
[WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
listen from /camserver/rgb
[Thread 0x7fffb77fe700 (LWP 21306) exited]
[Thread 0x7fffb6ffd700 (LWP 21307) exited]
[Thread 0x7fffb67fc700 (LWP 21308) exited]
[Thread 0x7fffb7fff700 (LWP 21305) exited]
lc-startyo[INFO] [camViz]: starting
Hello 
[Thread 0x7fffcb895700 (LWP 21299) exited]

Thread 1 "camViz" received signal SIGSEGV, Segmentation fault.
0x0000555555567265 in cv::Mat::release (this=0x7fffffffdd48)
    at /usr/include/opencv2/core/mat.inl.hpp:708
708	        size.p[i] = 0;
(gdb) backtrace
#0  0x0000555555567265 in cv::Mat::release() (this=0x7fffffffdd48)
    at /usr/include/opencv2/core/mat.inl.hpp:708
#1  0x0000555555566f3e in cv::Mat::~Mat() (this=0x7fffffffdd48, __in_chrg=<optimized out>) at /usr/include/opencv2/core/mat.inl.hpp:590
#2  0x000055555556801a in JdeRobotTypes::Image::~Image() (this=0x7fffffffdd18, __in_chrg=<optimized out>) at /opt/jderobot/include/jderobot/types/image.hpp:27
#3  0x00005555555691b2 in camViz::MinimalSubscriber::~MinimalSubscriber() (this=0x7fffffffdaa0, __in_chrg=<optimized out>)
    at /home/pankhuri/week8-camviz/camsubros2exp.hpp:66
#4  0x0000555555565f7a in main(int, char**) (argc=1, argv=0x7fffffffdc38)
    at /home/pankhuri/week8-camviz/camViz.cpp:81 ```

Useful links for similar errors:

1. https://answers.opencv.org/question/4072/initializing-static-cvmat-with-cvmatzeros-causes-segmentation-fault/

2.  https://github.com/opencv/opencv/issues/5174

3. https://answers.opencv.org/question/113449/how-to-initialize-mat-size-and-fill-it-with-zeros-in-a-class/

**Comment by pankhurivanjani** [comment]
Conclusion after debugging: This is not a JdeRobot Image library issue with ROS2. Cameraclient class is not returning the client and giving the segmentation fault

**Comment by pankhurivanjani** [comment]
CamViz -ROS2_cpp for review of errors:

Subscriber side:

Multithreaded spinner in Minimalsubscriber class (Listenercameraros2 kind of)

[spinner-camsubros2exp.hpp](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camsubros2exp.hpp#L36-L48)

Use of spinner in start function:

[spinner-start](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camsubros2exp.hpp#L78-L86)

Use of Minimal subscriber in getcameraclient class

[getcameraclient](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/mainsubcpp.hpp#L69-L95)

**Note: Error is coming in line [typecasting an instance of minimal subscriber class into client(instance of cameraclient class) ](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/mainsubcpp.hpp#L79)**

@vinay0410 I am working on this error taking care of your suggestions in the meeting. I have attached the snippets for your reference.

**Comment by pankhurivanjani** [comment]
Update: I was able to get rid of segmentation fault (I made minimalsubscriber inherit cameraclient), and the program was running without any errors but it seems the callbacks are getting blocked. 

I tested the program with a small script and conclusions are :

1. When I try to use multihreaded::executors' member function run:
  
* No errors but program not going into callback function ( might be blocking callbacks)

* Client in main file is able to access functions like getImage[get-image_function](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camsubros2exp.hpp#L162-L169) [2](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camViz.cpp#L66) without segmentation fault 

2. Instead of assigning Minimalsubscriber class to lc2 and then typecasting it into cameraclient class to return client, for debugging purpose when I run the program with spin:

``` rclcpp::init(argc, argv);
rclcpp::spin(std::make_shared<camViz::MinimalSubscriber>(0, nullptr, nodeName,topic));```

* Program does goes into callback function

* But we don't have anything to return to client this way, so we can't send anything to our GUI function.

So, We need the combined functionality both above methods. 

I will be trying to create my own multithreading function now, as suggested in the last meeting. 

@vinay0410 do you have any other suggestion for this, or shall I proceed with it?

**Comment by pankhurivanjani** [comment]
Update 2: Created a new branch for with a cleaner code for [Minimalsubscriber](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/ros2-callback_issue/camviz-updated/camsubros2exp.hpp).  Errors like segmentation fault have been solved, but the program is not going into [imagecallback](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/ros2-callback_issue/camviz-updated/camsubros2exp.hpp#L113-L132). I have tried what was suggested in forums as an alternative for asyncspinner but the only problem is the subscriber is not getting into callback function. 

There might be some problem in spinning it. One thing which I observed was the creation of multiple nodes by the same name. I corrected this too, but still not getting the right implementation.

Current Implementation:

1. [Minimalsubscriberclass]()

2. [call-class-in-ros2](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/ros2-callback_issue/camviz-updated/mainsubcpp.hpp#L69-L80)
